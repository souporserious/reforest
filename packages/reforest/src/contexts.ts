import * as React from "react"
import type { UseBoundStore, StoreApi } from "zustand"

export const PrerenderContext = React.createContext(false)

PrerenderContext.displayName = "PrerenderContext"

export const MaxIndexContext = React.createContext<number[]>([])

MaxIndexContext.displayName = "MaxIndexContext"

export const IndexContext = React.createContext<string | null>(null)

IndexContext.displayName = "IndexContext"

export type TreeState = {
  treeMap: Map<string, any>
  prerenderedTreeIds: Map<string, string>
  shouldPrerender: boolean
  setTreeData: (key: string, value: any, shouldUpdate?: boolean) => void
  deleteTreeData: (key: string, shouldUpdate?: boolean) => void
}

export type TreeStateStore = UseBoundStore<StoreApi<TreeState>>

export const TreeStateContext = React.createContext<TreeStateStore | null>(null)

TreeStateContext.displayName = "TreeStateContext"
