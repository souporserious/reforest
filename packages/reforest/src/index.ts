export type { TreeState, TreeStateStore } from "./contexts"
export { useIndex, useIndexedChildren } from "./use-indexed-children"
export { useRovingIndex } from "./use-roving-index"
export { useTree, useTreeId, useTreeNode, useTreeState, usePrerender } from "./use-tree"
export {
  cleanAndSortTree,
  compareIndexPaths,
  flattenChildren,
  mapToChildren,
  parseIndexPath,
  sortMapByIndexPath,
} from "./utils"
