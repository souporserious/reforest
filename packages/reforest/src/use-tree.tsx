import * as React from "react"
import { create } from "zustand"

import type { TreeState, TreeStateStore } from "./contexts"
import { PrerenderContext, TreeStateContext } from "./contexts"
import { useIndex, useIndexedChildren } from "./use-indexed-children"
import { sortMapByIndexPath, useIsomorphicLayoutEffect } from "./utils"

/**
 * Control tree state from outside a component.
 *
 * @example
 * import type { TreeMap } from "reforest"
 * import { useTree, useTreeNode, useTreeMap } from "reforest"
 *
 * function Item({ children, value }) {
 *   useTreeNode(value)
 *   return <li>{children}</li>
 * }
 *
 * function ItemList({ children }: { children: React.ReactNode, treeState: TreeState }) {
 *   const tree = useTree(children, treeState)
 *   return <ul>{tree.children}</ul>
 * }
 *
 * function App() {
 *   const treeState = useTreeState()
 *   return (
 *     <ItemList treeState={treeState}>
 *       <Item value="apple">Apple</Item>
 *       <Item value="banana">Banana</Item>
 *       <Item value="cherry">Cherry</Item>
 *     </ItemList>
 *   )
 * }
 */
export function useTreeState(): TreeStateStore
export function useTreeState<U>(selector: (state: TreeState) => U): U
export function useTreeState(selector?: (state: TreeState) => unknown) {
  const treeStateContext = React.useContext(TreeStateContext)
  const [treeState] = React.useState(
    () =>
      treeStateContext ||
      create<TreeState>((set, get) => ({
        treeMap: new Map(),
        prerenderedTreeIds: new Map(),
        shouldPrerender: true,
        setTreeData: (id, data) => {
          const { treeMap } = get()

          treeMap.set(id, data)

          set({ treeMap: sortMapByIndexPath(treeMap) })
        },
        deleteTreeData: (id) => {
          const { treeMap } = get()

          treeMap.delete(id)

          set({ treeMap: sortMapByIndexPath(treeMap) })
        },
      }))
  )

  return selector ? treeState(selector) : treeState
}

/** Prerenders children to capture data in useTreeNode hooks for initial component renders. */
function PrerenderTree({ children }: { children: React.ReactNode }) {
  const treeState = useTreeState()
  const shouldPrerender = treeState((state) => state.shouldPrerender)

  useIsomorphicLayoutEffect(() => {
    treeState.setState({
      prerenderedTreeIds: new Map(),
      shouldPrerender: false,
    })
  }, [])

  return shouldPrerender ? (
    <PrerenderContext.Provider value={true}>{children}</PrerenderContext.Provider>
  ) : null
}

/** Determine if the current render is a prerender. */
export function usePrerender() {
  const isPrerender = React.useContext(PrerenderContext)

  if (isPrerender === null) {
    throw new Error("usePrerender must be used in a descendant component of useTree.")
  }

  return isPrerender
}

/**
 * Manage ordered data subscriptions for components.
 *
 * @example create a tree of data subscriptions
 * import { useTree, useTreeNode } from "reforest"
 *
 * function Item({ children, value }) {
 *   useTreeNode(value)
 *   return <li>{children}</li>
 * }
 *
 * function ItemList({ children }: { children: React.ReactNode }) {
 *   const tree = useTree(children)
 *   return <ul>{tree.children}</ul>
 * }
 */
export function useTree(children: React.ReactNode, treeState?: TreeStateStore) {
  const treeStateContext = React.useContext(TreeStateContext)
  const treeStateLocal = useTreeState()
  const parsedTreeState = treeStateContext || treeState || treeStateLocal
  const isPrerender = React.useContext(PrerenderContext)
  const isRoot = treeStateContext === null
  const indexedChildren = useIndexedChildren(children)
  const childrenToRender = isRoot ? (
    <TreeStateContext.Provider value={parsedTreeState}>
      <PrerenderTree>{indexedChildren}</PrerenderTree>
      {indexedChildren}
    </TreeStateContext.Provider>
  ) : (
    indexedChildren
  )

  return {
    children: childrenToRender,
    useStore: parsedTreeState,
    isPrerender,
    isRoot,
  }
}

/** Generate an id for use with useTreeNode. */
export function useTreeId() {
  const treeStateContext = React.useContext(TreeStateContext)

  if (treeStateContext === null) {
    throw new Error("useTreeId must be used in a descendant component of useTree.")
  }

  const { prerenderedTreeIds } = treeStateContext.getState()
  const { indexPathString } = useIndex()!
  const generatedId = React.useId().slice(1, -1)
  const treeId = prerenderedTreeIds.get(indexPathString) || generatedId

  return treeId
}

/** Subscribe data to the root useTree hook. */
export function useTreeNode(
  treeId: string,
  getData: () => any,
  dependencies: React.DependencyList = []
) {
  const isPrerender = React.useContext(PrerenderContext)
  const treeStateContext = React.useContext(TreeStateContext)

  if (treeStateContext === null) {
    throw new Error("useTreeNode must be used in a descendant component of useTree.")
  }

  const { deleteTreeData, prerenderedTreeIds, setTreeData, treeMap } = treeStateContext.getState()
  const { indexPathString } = useIndex()!
  const treeData = React.useMemo(
    () => Object.assign({ treeId }, getData()),
    dependencies.concat(treeId)
  )

  if (isPrerender) {
    /** Mutate tree data when prerendering so it's available when doing the subsequent render of root children. */
    treeMap.set(indexPathString, treeData)

    /** Store the treeId so there's a stable id between prerender and actual render. */
    prerenderedTreeIds.set(indexPathString, treeId)
  } else {
    /** After the initial prerender we switch to a simple effect for coordinating data updates. */
    React.useEffect(() => {
      setTreeData(indexPathString, treeData)

      return () => {
        deleteTreeData(indexPathString)
      }
    }, [indexPathString, treeData])
  }

  return treeData
}
