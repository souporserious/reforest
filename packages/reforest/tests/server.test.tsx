/** @jest-environment node */
import * as React from "react"
import { renderToString } from "react-dom/server"

import { useTree, useTreeId, useTreeNode, useTreeState, usePrerender } from "../src"

test("computed data renders on server", () => {
  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const tree = useTree(children)
    const treeId = useTreeId()
    const isPrerender = usePrerender()

    useTreeNode(treeId, () => ({ value }), [value])

    if (isPrerender) {
      return null
    }

    const treeMap = useTreeState((state) => state.treeMap)

    return (
      <div data-testid={treeId}>
        {treeMap.size} {tree.children}
      </div>
    )
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    return tree.children
  }

  const renderedString = renderToString(
    <ItemList>
      <Item value="apple">Apple</Item>
      <Item value="orange">Orange</Item>
      <Item value="banana">Banana</Item>
    </ItemList>
  )

  expect(renderedString).toMatchSnapshot()
})

test("changing rendered elements based on computed data", () => {
  function Box({ id }: { id: string }) {
    const treeId = useTreeId()
    const isPrerender = usePrerender()

    useTreeNode(treeId, () => ({ id }), [id])

    if (isPrerender) {
      return null
    }

    const treeMap = useTreeState((state) => state.treeMap)
    const ids = new Set()
    let shouldRender = false

    treeMap.forEach((treeNode) => {
      const isSameInstance = treeNode.treeId === treeId
      const hasId = ids.has(treeNode.id)

      if (isSameInstance) {
        shouldRender = !hasId
      }

      if (!hasId) {
        ids.add(treeNode.id)
      }
    })

    return shouldRender ? <div id={id} /> : null
  }

  function Parent({ children }: { children: React.ReactNode }) {
    const tree = useTree(children)

    return tree.children
  }

  const renderedString = renderToString(
    <Parent>
      <Box id="a" />
      <Box id="b" />
      <Box id="c" />
      <Box id="b" />
    </Parent>
  )

  expect(renderedString).toMatchSnapshot()
})
