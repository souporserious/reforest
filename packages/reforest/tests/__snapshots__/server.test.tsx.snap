// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changing rendered elements based on computed data 1`] = `"<div id="a"></div><div id="b"></div><div id="c"></div>"`;

exports[`computed data renders on server 1`] = `"<div data-testid="R5">3<!-- --> <!-- -->Apple</div><div data-testid="R9">3<!-- --> <!-- -->Orange</div><div data-testid="Rd">3<!-- --> <!-- -->Banana</div>"`;
